stumpy/scrump.py

```diff
@@ -714,7 +718,11 @@ def __init__(
             else:
                 P, I = prescrump(T_A, m, T_B=T_B, s=s, k=self._k)
 
-            core._merge_topk_PI(self._P, P, self._I, I)
+            # P and I are 1D when `self._k` is 1. So, we should reshape them
+            # before passing them to `_merge_topk_PI`
+            core._merge_topk_PI(
+                self._P, P.reshape(-1, self._k), self._I, I.reshape(-1, self._k)
+            )
```

Comment:
I think that when `k=1`, it makes sense for `self._P` and `self._I` to also be 1D. You can accomplish this with `np.squeeze`:

```
self._P = np.full((self._l, self._k), np.inf, dtype=np.float64).squeeze()
```

`squeeze` automatically removes any axes of length one

Reply:
That should work. However, we then have a slight problem in `.update()`

```
# lines 763-783
P, PL, PR, I, IL, IR = _stump(
                ...
            )
            # Update (top-k) matrix profile and indices
            core._merge_topk_PI(self._P, P, self._I, I)
```
Here, `self._P` and `self._I` are 1D, but `P` and `I` (from `_stump`) are 2D. Do you think I should make `P` and `I` 1D for `k=1` in `_stump`?

Reply:
> It seems that you'd want to check that the input arrays are of the same dimensions though (i.e., both arrays to be merged are 1D OR both arrays are 2D)

I am thinking of explaining this in the docstring that inputs must have the same shape (all 1D or all 2D), but I am not going to add a code to check it. I think we cannot raise error since we are using it in `@njit`-decorated functions.

Reply:
> I need to reshape P and I to 2D when k=1.

> P and I (from _stump) are 2D

I am confused now.

Reply:
This is what I see in my head...

if we go with:
```
self._P = np.full((l, self._k), np.inf, dtype=np.float64).squeeze()
self._I = np.full((l, self._k), np.inf, dtype=np.float64).squeeze()
```
Then, `self._P` and `self._I` are both 1D when `k=1`. So far so good... However, in `.update()`, we can see that we need to do

```
core._merge_topk_PI(self._P, P, self._I, I)  
```

to update `self._P` and `self._I`. However, the `P` and `I`,  obtained from `_stump`, are 2D. In `_stump`, output is ALWAYS 2D, even if `k=1`.

So, I think I should go with what you suggested in another comment: using `.squeeze()` trick in `_merge_topk_PI`. I can do:

```
def _merge_topk_PI(PA, PB, IA, IB):
     PA = np.squeeze(PA)
     PB = np.squeeze(PB)
     IA = np.squeeze(IA)
     IB = np.squeeze(IB)
     
     if PA.ndim == 1:
          # do 1D merge
     else:
          k = PA.shape[1]
          # do 2D merge
```

If I missed something or my explanation was not clear, please let me know.

Reply:
```
def _merge_topk_PI(PA, PB, IA, IB):
     PA = np.squeeze(PA)
     PB = np.squeeze(PB)
     IA = np.squeeze(IA)
     IB = np.squeeze(IB)
```

I don't think this code will work though because you are making a copy of `PA`, `PB`, `IA`, and `IB` inside of the function and those arrays do not get returned and used. I think the `squeeze` needs to happen outside of `_merge_topk_PI`

> core._merge_topk_PI(self._P, P, self._I, I)  

Hmmm, maybe I'm going back and forth here. Sorry for that. Perhaps, the best thing to do is to always try to keep everything as 2D and then convert to 1D in the end or when we access `self.P_` and `self.I_`? In that case, instead of reshaping something from 1D to 2D when `k=1`, it is probably more obvious to use `np.atleast_2d`:

```
np.atleast_2d(np.full(10, np.inf))
```

This will ensure that any array is at least 2D. So, 1D arrays will get converted to 2D and 2D arrays will be left untouched. I just don't love `reshape` as it is more opaque and non obvious what is happening

Reply:
> I don't think this code will work though because you are making a copy of `PA`, `PB`, `IA`, and `IB` inside of the function and those arrays do not get returned and used.

I thought `PA = np.squeeze(PA)` creates a view and hence changing PA on the left inside the function changes the PA on the right. Now, I can see that I am over-writing the `PA`. So, maybe(?) `PAA = np.squeeze(PA)` works. (I have not tried it out. So, I am not sure.)

----

> Hmmm, maybe I'm going back and forth here. Sorry for that.

I see it as a learning process. All good :)


> I'm guessing that this is rare?

Yeah...I think so. However, in the future development of the library, one should be careful when writing a code. If they reuse the code or some of these functions, they should take care of 1D-vs-2D cases. So, I think this suggestion `"Perhaps, the best thing to do is to always try to keep everything as 2D and then convert to 1D in the end"` might be better than `"if we are forced to call a function from within another function (i.e., we aren't calling a function that returns to the user) and it returns 1D but we need 2D, then we should probably convert it to 2D"`. In the latter, we are first making them 1D, and then if we need it 2D, we should convert it back to 2D. I think the former suggestion, i.e. keeping everything 2D and make it 1D in the end, might be easier for future development and add less complexity to the code.

Also, if we rename the prescrump function as follows:
```
_prescrump ---> _computing_prescrump
prescrump ---> _prescrump # this always returns 2D
prescrump ---> simple wrapper for _prescrump, that makes the output 1D for user
```

Then, we can call this new `_prescrump` inside scrump and everything will be good I guess.